src/types/utils.ts

```diff
@@ -158,51 +158,80 @@ export type Not<T extends boolean> = T extends true ? false : true;
  * Union of the sets' values' types
  */
 export type UnionSetValues<Ts extends ReadonlyArray<unknown>> =
-  Ts extends readonly [infer Head, ...infer Rest]
-    ? Head extends Set<infer V1>
-      ? Rest extends ReadonlyArray<unknown>
-        ? UnionSetValues<Rest> | V1
-        : V1
-      : never
-    : never;
+  UnionSetValuesHelper<Ts, never>;
+
+/**
+ * Tail-recursive helper type for UnionSetValues.
+ */
+type UnionSetValuesHelper<
+  Ts extends ReadonlyArray<unknown>,
+  Acc
+> = Ts extends readonly [infer Head, ...infer Rest]
+  ? Head extends Set<infer V1>
+    ? Rest extends ReadonlyArray<unknown>
+      ? UnionSetValuesHelper<Rest, Acc | V1>
+      : Acc | V1
+    : never
+  : Acc;
 
 /**
  * Union of the maps' values' types
  */
 export type UnionMapKeys<Ts extends ReadonlyArray<unknown>> =
-  Ts extends readonly [infer Head, ...infer Rest]
-    ? Head extends Map<infer K1, unknown>
-      ? Rest extends readonly []
-        ? K1
-        : K1 | UnionMapKeys<Rest>
-      : never
-    : never;
+  UnionMapKeysHelper<Ts, never>;
+
+/**
+ * Tail-recursive helper type for UnionMapKeys.
+ */
+type UnionMapKeysHelper<
+  Ts extends ReadonlyArray<unknown>,
+  Acc
+> = Ts extends readonly [infer Head, ...infer Rest]
+  ? Head extends Map<infer K1, unknown>
+    ? Rest extends readonly []
+      ? Acc | K1
+      : UnionMapKeysHelper<Rest, Acc | K1>
+    : never
+  : Acc;
 
 /**
  * Union of the maps' keys' types
  */
 export type UnionMapValues<Ts extends ReadonlyArray<unknown>> =
-  Ts extends readonly [infer Head, ...infer Rest]
-    ? Head extends Map<unknown, infer V1>
-      ? Rest extends readonly []
-        ? V1
-        : UnionMapValues<Rest> | V1
-      : never
-    : never;
+  UnionMapValuesHelper<Ts, never>;
+
+/**
+ * Tail-recursive helper type for UnionMapValues.
+ */
+type UnionMapValuesHelper<
+  Ts extends ReadonlyArray<unknown>,
+  Acc
+> = Ts extends readonly [infer Head, ...infer Rest]
+  ? Head extends Map<unknown, infer V1>
+    ? Rest extends readonly []
+      ? Acc | V1
+      : UnionMapValuesHelper<Rest, Acc | V1>
+    : never
+  : Acc;
 
 /**
  * Get all the keys of the given records.
  */
-export type KeysOf<Ts extends ReadonlyArray<unknown>> = Ts extends readonly [
-  infer Head,
-  ...infer Rest
-]
+export type KeysOf<Ts extends ReadonlyArray<unknown>> = KeysOfHelper<Ts, never>;
+
+/**
+ * Tail-recursive helper type for KeysOf.
+ */
+type KeysOfHelper<
+  Ts extends ReadonlyArray<unknown>,
+  Acc
+> = Ts extends readonly [infer Head, ...infer Rest]
   ? Head extends Record<RecordProperty, unknown>
```

Comment:
You can use built in `PropertyKey` type instead of `RecordProperty` which is `keyof any`